File: pocs/stack/java/src/main/java/stack/Stack.java
package stack;

import java.util.EmptyStackException;
import java.util.Iterator;
import java.util.NoSuchElementException;

/**
 * Generic Stack implementation using a singly linked list.
 * LIFO (Last In, First Out) data structure.
 *
 * Time Complexity:
 *   push  -> O(1)
 *   pop   -> O(1)
 *   peek  -> O(1)
 *   size  -> O(1)
 *
 * Space Complexity: O(n)
 *
 * @param <T> the type of elements held in this stack
 */
public class Stack<T> implements Iterable<T> {
    private static class Node<T> {
        T data;
        Node<T> next;

        Node(T data) {
            this.data = data;
        }
    }

    private Node<T> top;
    private int size;


    /**
     * Pushes an element onto the top of the stack.
     * @param item element to push
     * @throws IllegalArgumentException if item is null
     */
    public void push(T item) {
        if (item == null) throw new IllegalArgumentException("Null elements are not allowed");
        Node<T> newNode = new Node<>(item);
        newNode.next = top;
        top = newNode;
        size++;
    }

    /**
     * Removes and returns the element at the top of the stack.
     * @return the top element
     * @throws EmptyStackException if the stack is empty
     */
    public T pop() {
        if (isEmpty()) throw new EmptyStackException();
        T data = top.data;
        top = top.next;
        size--;
        return data;
    }

    /**
     * Returns (without removing) the element at the top of the stack.
     * @return the top element
     * @throws EmptyStackException if the stack is empty
     */
    public T peek() {
        if (isEmpty()) throw new EmptyStackException();
        return top.data;
    }

    /**
     * Returns true if the stack contains no elements.
     */
    public boolean isEmpty() {
        return size == 0;
    }

    /**
     * Returns the number of elements in the stack.
     */
    public int size() {
        return size;
    }

    /**
     * Removes all elements from the stack.
     */
    public void clear() {
        top = null;
        size = 0;
    }

    /**
     * Returns true if the stack contains the specified element.
     * O(n) linear scan.
     */
    public boolean contains(T item) {
        Node<T> current = top;
        while (current != null) {
            if (current.data.equals(item)) return true;
            current = current.next;
        }
        return false;
    }

    @Override
    public Iterator<T> iterator() {
        return new Iterator<>() {
            private Node<T> current = top;

            @Override
            public boolean hasNext() {
                return current != null;
            }

            @Override
            public T next() {
                if (!hasNext()) throw new NoSuchElementException();
                T data = current.data;
                current = current.next;
                return data;
            }
        };
    }

    @Override
    public String toString() {
        if (isEmpty()) return "Stack[]";
        StringBuilder sb = new StringBuilder("Stack[top -> ");
        Node<T> current = top;
        while (current != null) {
            sb.append(current.data);
            if (current.next != null) sb.append(", ");
            current = current.next;
        }
        return sb.append("]").toString();
    }
}

File: pocs/stack/typescript/src/stack.ts
/**
 * Generic Stack implementation using a singly linked list.
 * LIFO (Last In, First Out) data structure.
 *
 * Time Complexity:
 *   push  -> O(1)
 *   pop   -> O(1)
 *   peek  -> O(1)
 *   size  -> O(1)
 *
 * Space Complexity: O(n)
 */
interface Node<T> {
  data: T;
  next: Node<T> | null;
}

export class EmptyStackError extends Error {
  constructor() {
    super("Stack is empty");
    this.name = "EmptyStackError";
  }
}

export class Stack<T> {
  private top: Node<T> | null = null;
  private _size: number = 0;

  /**
   * Pushes an element onto the top of the stack.
   * @throws {TypeError} if item is null or undefined
   */
  push(item: T): void {
    if (item === null || item === undefined) {
      throw new TypeError("Null or undefined elements are not allowed");
    }
    const newNode: Node<T> = { data: item, next: this.top };
    this.top = newNode;
    this._size++;
  }

  /**
   * Removes and returns the element at the top of the stack.
   * @throws {EmptyStackError} if the stack is empty
   */
  pop(): T {
    if (this.isEmpty()) throw new EmptyStackError();
    const data = this.top!.data;
    this.top = this.top!.next;
    this._size--;
    return data;
  }

  /**
   * Returns (without removing) the element at the top of the stack.
   * @throws {EmptyStackError} if the stack is empty
   */
  peek(): T {
    if (this.isEmpty()) throw new EmptyStackError();
    return this.top!.data;
  }

  /** Returns true if the stack contains no elements. */
  isEmpty(): boolean {
    return this._size === 0;
  }

  /** Returns the number of elements in the stack. */
  get size(): number {
    return this._size;
  }

  /** Removes all elements from the stack. */
  clear(): void {
    this.top = null;
    this._size = 0;
  }

  /**
   * Returns true if the stack contains the given item.
   * Uses strict equality (===). O(n).
   */
  contains(item: T): boolean {
    let current = this.top;
    while (current !== null) {
      if (current.data === item) return true;
      current = current.next;
    }
    return false;
  }

  [Symbol.iterator](): Iterator<T> {
    let current = this.top;
    return {
      next(): IteratorResult<T> {
        if (current !== null) {
          const value = current.data;
          current = current.next;
          return { value, done: false };
        }
        return { value: undefined as unknown as T, done: true };
      },
    };
  }

  /** Returns an array snapshot from top to bottom. */
  toArray(): T[] {
    const result: T[] = [];
    for (const item of this) {
      result.push(item);
    }
    return result;
  }

  toString(): string {
    if (this.isEmpty()) return "Stack[]";
    return `Stack[top -> ${this.toArray().join(", ")}]`;
  }
}
